src/Validation/Rule/VariablesInAllowedPositionRule.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace Digia\GraphQL\Validation\Rule;
+
+use Digia\GraphQL\Error\InvalidTypeException;
+use Digia\GraphQL\Error\ValidationException;
+use Digia\GraphQL\Language\Node\NodeInterface;
+use Digia\GraphQL\Language\Node\OperationDefinitionNode;
+use Digia\GraphQL\Language\Node\VariableDefinitionNode;
+use Digia\GraphQL\Language\Node\VariableNode;
+use Digia\GraphQL\Type\Definition\NonNullType;
+use Digia\GraphQL\Type\Definition\TypeInterface;
+use Digia\GraphQL\Util\TypeComparator;
+use function Digia\GraphQL\Type\GraphQLNonNull;
+use function Digia\GraphQL\Util\typeFromAST;
+use function Digia\GraphQL\Validation\badVariablePositionMessage;
+
+/**
+ * Variables in allowed position
+ *
+ * Variables passed to field arguments conform to type.
+ */
+class VariablesInAllowedPositionRule extends AbstractRule
+{
+    /**
+     * @var VariableDefinitionNode[]|null
+     */
+    protected $variableDefinitionMap;
+
+    /**
+     * @var TypeComparator
+     */
+    protected $typeComparator;
+
+    /**
+     * VariablesInAllowedPositionRule constructor.
+     */
+    public function __construct()
+    {
+        $this->typeComparator = new TypeComparator();
```

Comment:
This is obviously not ideal but we can probably come up with a solution later.

Reply:
Yes, but this is the middle ground.